### PR TITLE
Add Entra Id authentication support for installer download

### PR DESCRIPTION
**Pull request by yao-msft**
- The entra id support for installer download reuses existing entra id code for rest source access (parsing, authenticating).
- Added yaml manifest parsing for authentication to make manifest in parity with rest source. Added validation to disable Authentication in manifest for community repo.
- Validated manually with private Azure blob storage hosted installer.
- Added manifest tests and installer download authentication unit tests
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-cli/pull/5095)

### CODE REVIEW

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 1]
```diff
@@ -3187,4 +3187,14 @@ Please specify one of them using the --source option to proceed.</value>
     <value>Configuration Modules</value>
     <comment>PowerShell Modules that are used for the Configuration feature</comment>
   </data>
-</root>
\ No newline at end of file
+  <data name="InstallerDownloadRequiresAuthentication" xml:space="preserve">
+    <value>The package installer requires authentication. Authentication prompt may appear when necessary. Authenticated information will be shared with the installer download url.</value>
+  </data>
+  <data name="InstallerDownloadAuthenticationNotSupported" xml:space="preserve">
+    <value>Failed to download installer. This winget version does not support the installer download authentication method. Try upgrade to latest winget version.</value>
```
Comment (ryfu-msft) [inline src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:3194]:
```suggestion
    <value>Failed to download installer. This winget version does not support the installer download authentication method. Try upgrading to latest winget version.</value>

```

File: src/AppInstallerCommonCore/Authentication/Authentication.cpp [hunk 2]
```diff
@@ -54,12 +56,48 @@ namespace AppInstaller::Authentication
         THROW_HR_IF(E_UNEXPECTED, !m_authProvider);
 
         return m_authProvider->AuthenticateForToken();
+    }
+
+    bool MicrosoftEntraIdAuthenticationInfo::operator<(const MicrosoftEntraIdAuthenticationInfo& other) const
+    {
+        // std::tie implements tuple comparison, wherein it checks the first item in the tuple,
+        // iff the first elements are equal, then the second element is used for comparison, and so on
```
Comment (ryfu-msft) [inline src/AppInstallerCommonCore/Authentication/Authentication.cpp:64]:
```suggestion
        // if the first elements are equal, then the second element is used for comparison, and so on

```

File: src/AppInstallerCommonCore/Authentication/Authentication.cpp [hunk 3]
```diff
@@ -54,12 +56,48 @@ namespace AppInstaller::Authentication
...
+    {
+        // std::tie implements tuple comparison, wherein it checks the first item in the tuple,
+        // iff the first elements are equal, then the second element is used for comparison, and so on
+        return std::tie(Resource, Scope) < std::tie(other.Resource, other.Scope);
+    }
+
+    bool AuthenticationInfo::operator<(const AuthenticationInfo& other) const
+    {
+        // std::tie implements tuple comparison, wherein it checks the first item in the tuple,
+        // iff the first elements are equal, then the second element is used for comparison, and so on
```
Comment (ryfu-msft) [inline src/AppInstallerCommonCore/Authentication/Authentication.cpp:71]:
```suggestion
        // if the first elements are equal, then the second element is used for comparison, and so on

```

File: src/AppInstallerCommonCore/Authentication/Authentication.cpp [hunk 4]
```diff
@@ -24,9 +26,9 @@ namespace AppInstaller::Authentication
 
         AICLI_LOG(Core, Info, << "AuthenticationArguments values. Mode: " << AuthenticationModeToString(args.Mode) << ", Account: " << args.AuthenticationAccount);
 
-        if (info.Type == AuthenticationType::MicrosoftEntraId)
+        if (info.Type == AuthenticationType::MicrosoftEntraId || info.Type == AuthenticationType::MicrosoftEntraIdForAzureBlobStorage)
```
Comment (ryfu-msft) [inline src/AppInstallerCommonCore/Authentication/Authentication.cpp:29]: nit: Should we create a semantic function for this similar to what you did for WebAccountAuthenticator?

**Review by ryfu-msft — APPROVED** [review]

**Review by ryfu-msft — APPROVED** [review]
:shipit:

**Comment by Elonmusk678** [comment]
> :shipit:

**Comment by Elonmusk678** [comment]
> :shipit: 

How are you

**Pull request merged**